bin/main.ml

```diff
@@ -749,6 +788,56 @@ let resolve_targets_exn ~log common setup user_targets =
     | Ok targets ->
       targets)
 
+let fswatch_command root_path =
+  let excludes = [ {|\.#|}
+                 ; {|_build|}
+                 ; {|\.hg|}
+                 ; {|\.git|}
+                 ; {|~$|}
+                 ; {|/#[^#]*#$|}
+                 ; {|\.install$|}
+                 ]
+  in
+  let path = Path.to_string_maybe_quoted root_path in
+  let uname =
+    match Bin.which "uname" with
+    | Some uname ->
+      Process.run_capture Strict uname [] ~env:Env.initial >>| String.trim
+    | None -> Fiber.return "Unknown"
+  in
+  uname >>| function
+  | "Linux" ->
+    (* On Linux, use inotifywait. *)
+    let excludes = String.concat ~sep:"|" excludes in
+    "inotifywait",
+    ["-r"; path; "--exclude"; excludes; "-e"; "close_write"; "-q"]
+  | _ ->
+    (* On all other platforms, try to use fswatch. fswatch's event
+       filtering is not reliable (at least on Linux), so don't try to
+       use it, instead act on all events. *)
+    let excludes =
+      List.(map ~f:(fun x -> ["--exclude"; x]) excludes |> concat)
+    in
+    let args = [ "-r"; path; "-1" ] @ excludes in
+    "fswatch", args
+
+let clear_all_caches () =
+  Dir_contents.clear_cache ();
+  Report_error.clear_cache ()
+
+let watch_changes () =
+  fswatch_command Path.root
+  >>= fun (fswatch, args) ->
+  let fswatch =
+    match Bin.which fswatch with
+    | Some x -> x
+    | None ->
+      (* Don't throw an exception here to prevent a loop. *)
+      Format.eprintf "Error: executable not found: %s\n" fswatch;
+      exit 1
+  in
+  Process.run_capture Strict fswatch args ~env:Env.initial
```

Comment:
I'm not sure that this is intended, but it seems that all call sites of `watch_changes` pipe it into `ignore`, so it might be worth returning `unit` directly instead.

Reply:
Or even: `Process.run ~stdout:Config.dev_null`

Reply:
I've added `>>| ignore` for now, but in the future we're likely to need the actual names of files that were modified, so this is going to change anyway.